### PR TITLE
Move to using custom Ingestor for cloud resource specifications

### PR DESCRIPTION
**Pull request by eiffel777**
Move to using a custom Ingestor class for setting the start and end time for a host configuration in a cloud resource

## Description
The custom Ingestor makes it easier to track when a host for a cloud resource changes the number of vcps or memory it contains and when a host may be taken out of a cloud resource.

## Motivation and Context
This changes eliminates bugs in the previous implementation and makes it clearer to understand what is going on.

## Tests performed
Tested in docker with fabricated test data and add new component tests

## Checklist:

- [x] The pull request description is suitable for a Changelog entry
- [x] The milestone is set correctly on the pull request
- [x] The appropriate labels have been added to the pull request

### CODE REVIEW

File: classes/ETL/Ingestor/CloudResourceSpecsStateTransformIngestor.php [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        // Since we only get information for when a configuration changes we assume a configuration has an end date
+        // of today unless we have a row that tells us otherwise
+        $default_end_time = isset($this->_end_time) ? $this->_end_time : date('Y-m-d') . ' 23:59:59';
+
+        $this->_instance_state = array(
+            'resource_id' => $srcRecord['resource_id'],
+            'hostname' => $srcRecord['hostname'],
+            'vcpus' => $srcRecord['vcpus'],
+            'memory_mb' => $srcRecord['memory_mb'],
+            'start_date_ts' => strtotime($srcRecord['start_date_ts'] . " 00:00:00"),
```
Comment (jpwhite4) [inline classes/ETL/Ingestor/CloudResourceSpecsStateTransformIngestor.php:52]: We conventionally use the _ts postfix to indicate that the data type is a posix timestamp. This looks like it is converting a datetime to a timestamp in place. Is this intended?

Comment (eiffel777) [reply classes/ETL/Ingestor/CloudResourceSpecsStateTransformIngestor.php]: @jpwhite4 Yeah, it is a datetime being converted to a timestamp. It's just wasn't named correctly. I've changed the name to denote it is a date

**Comment by eiffel777** [comment]
@jpwhite4 @ryanrath This is updated and ready to be reviewed

**Comment by jpwhite4** [comment]
This code does not generate correct SQL:

#10 {main}","message":"xdmod.aggregate-cloud-resource-specs.CloudResourceSpecsAggregator (ETL\\Aggregator\\SimpleAggregator): Error processing aggregation period Exception: 'SQLSTATE[42000]: Syntax error or access violation: 1055 'modw_cloud.crs.vcpus' isn't in GROUP BY'"}

In your tests I suggest merging #1469 so that the framework reports the errors.

File: classes/ETL/Ingestor/CloudResourceSpecsStateTransformIngestor.php [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        if ($srcRecord['fact_date'] === 0) {
+            if (isset($this->_instance_state)) {
+                return array($this->_instance_state);
+            } else {
+                return array();
+            }
+        }
+
+        if ($this->_instance_state === null) {
+            if($srcRecord['vcpus'] == -1 && $srcRecord['memory_mb'] == -1) {
```
Comment (jpwhite4) [inline classes/ETL/Ingestor/CloudResourceSpecsStateTransformIngestor.php:89]: What scenario is handled here? When would you have the first record have vcpus and memory_mb as -1?

**Review by jpwhite4 — APPROVED** [review]

**Pull request merged**